File: js/src/cpu-async/concurrent/concurrent.ts
/* eslint-disable no-console */
const crypto = (await import('crypto')).default;

export function load() {
  return new Promise<void>((resolve, reject) => {
    const salt =
      'Gg8au1BBLgE1MGQGiRULunMqOCoJvhHA3qXhr3FFn+Z7iWiLMmNI6j+JIcd6ledP';
    return crypto.pbkdf2('myPassword', salt, 50000, 512, 'sha512', (err, _key) => {
        if (err) reject(err);
        else resolve();
      },
    );
  });
}

function routine(isCancelled: () => boolean, job: number) {
  if (isCancelled()) {
    console.log('cancelled: ', job);
    return Promise.resolve();
  }
  console.log('work: ', job);
  return load().then(() => console.log('done: ', job));
}

function processor(
  isCancelled: () => boolean,
  _cancel: () => void,
  jobs: number[],
) {
  console.time('processor');
  let promises: Promise<void>[] = [];

  jobs.forEach((job, i) => {
    const promise = routine(isCancelled, job);
    promises.push(promise);
  });

  return Promise.all(promises).then(() => console.timeEnd('processor'));
}

export async function main() {
  const jobs = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17];
  let _cancelled = false;
  const cancel = () => {
    _cancelled = true;
  };
  const getCancel = () => _cancelled;
  await processor(getCancel, cancel, jobs);
}

main();

File: go/cpu/concurrent/concurrent.go
package concurrent

import (
	"context"
	"crypto/sha512"
	"fmt"
	"sync"
	"time"

	"golang.org/x/crypto/pbkdf2"
)

const CONCURENCY_LIMIT = 10

var wg sync.WaitGroup

// bchannel is a bounded channel allowing up to 10 concurrent operations.
var bchannel = make(chan int, CONCURENCY_LIMIT)

func load() {
	salt := []byte("Gg8au1BBLgE1MGQGiRULunMqOCoJvhHA3qXhr3FFn+Z7iWiLMmNI6j+JIcd6ledP")
	pbkdf2.Key([]byte("myPassword"), salt, 50000, 512, sha512.New)
}

func routine(ctx context.Context, job interface{}) {

	select {
	case <-ctx.Done():
		fmt.Printf("%#v: cancelled\n", job)
	default:
		fmt.Printf("%#v: work\n", job)
		// do work
		load()
		fmt.Printf("%#v: done\n", job)
	}

	<-bchannel // unblock loop
	wg.Done()  // unblock main
}

func Processor(ctx context.Context, cancel context.CancelFunc, jobs []int) {
	t1 := time.Now()
	for _, job := range jobs {
		bchannel <- 1 // blocks loop if len(bchannel) == CONCURENCY_LIMIT
		wg.Add(1)     // blocks main
		go routine(ctx, job)
	}

	wg.Wait()
	elapsed := time.Since(t1).String()
	fmt.Printf("processed %d jobs in %s\n", len(jobs), elapsed)
}
